// @repositive/iris/dist/amqp/index.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const amqplib_1 = require("amqplib");
const register_1 = require("./register");
const request_1 = require("./request");
function restartConnection({ opts, timeout = 100, attempt = 0, _setup = setup, _setTimeout = setTimeout }) {
    return new Promise((resolve, reject) => {
        const _log = opts._log || console;
        const _restartConnection = opts._restartConnection || restartConnection;
        _log.info(`Retrying connection in ${attempt * timeout}ms`);
        _setTimeout(() => {
            resolve(_setup(opts).catch((innerErr) => {
                return _restartConnection({ opts, timeout, _setup, attempt: attempt <= 100 ? attempt + 10 : attempt });
            }));
        }, attempt * timeout);
    });
}
exports.restartConnection = restartConnection;
const defaults = {
    uri: 'amqp://guest:guest@localhost',
    exchange: 'iris',
    namespace: 'default',
    registrations: {},
    _setupRequest: request_1.setupRequest,
    _setupRegister: register_1.setupRegister,
    _connect: amqplib_1.connect,
    _restartConnection: restartConnection,
    _log: console
};
function setup(opts = defaults) {
    return __awaiter(this, void 0, void 0, function* () {
        const _opts = Object.assign({}, defaults, opts);
        const { uri, exchange, registrations, _setupRequest, _setupRegister, _connect, _restartConnection, _log } = _opts;
        const common_options = { durable: true, noAck: true };
        const conn = yield _connect(uri, common_options);
        const channel = yield conn.createChannel();
        const options = Object.assign({ ch: channel }, _opts);
        const operations = yield Promise.all([
            _setupRequest(options),
            _setupRegister(options)
        ]);
        let errored = false;
        function onError(error) {
            if (!errored) {
                errored = true;
                _log.warn(`Connection errored...`);
                _restartConnection({ opts: _opts }).then(({ register, request }) => {
                    operations[0] = request;
                    operations[1] = register;
                    errored = false;
                    _log.info('Connection recovered');
                })
                    .catch((err) => {
                    _log.error(err);
                    /* This promise should never reject */
                });
            }
        }
        conn.on('error', onError);
        conn.on('close', onError);
        // If the connection failed there may be subscriptions from previous connection, so add them again.
        yield Promise.all(Object.keys(registrations).map(k => {
            const registration = registrations[k];
            return operations[1](registration);
        }));
        return {
            register(ropts) {
                return __awaiter(this, void 0, void 0, function* () {
                    const id = `${ropts.pattern}-${ropts.namespace || ''}`;
                    if (!registrations[id]) {
                        registrations[id] = ropts;
                    }
                    if (errored) {
                        return Promise.resolve();
                    }
                    else {
                        return operations[1](ropts);
                    }
                });
            },
            request(ropts) {
                return __awaiter(this, void 0, void 0, function* () {
                    if (errored) {
                        return Promise.reject(new Error('Broken pipe'));
                    }
                    else {
                        return operations[0](ropts);
                    }
                });
            }
        };
    });
}
exports.default = setup;
