// @repositive/iris/dist/amqp/index.spec.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const test = require("tape");
const sinon_1 = require("sinon");
const index_1 = require("./index");
const index_2 = require("./index");
const _restartConnection = sinon_1.spy();
function mockConnect() {
    return {
        createChannel: sinon_1.spy(),
        on: sinon_1.spy()
    };
}
function wait(time) {
    return new Promise((resolve, reject) => {
        setTimeout(() => resolve(), time);
    });
}
function mockOpts() {
    const connectResponse = mockConnect();
    const register = sinon_1.spy();
    const request = sinon_1.spy();
    const registrations = { 'test': { pattern: 'test', handler: sinon_1.spy() } };
    return {
        steps: {
            request,
            register,
            connectResponse
        },
        mocks: {
            uri: '',
            exchange: '',
            registrations,
            _setupRequest: sinon_1.stub().returns(Promise.resolve(request)),
            _setupRegister: sinon_1.stub().returns(Promise.resolve(register)),
            _restartConnection: sinon_1.stub().returns(Promise.resolve({ request, register })),
            _connect: sinon_1.stub().returns(Promise.resolve(connectResponse)),
            _log: { log: sinon_1.spy(), info: sinon_1.spy(), warn: sinon_1.spy(), error: sinon_1.spy() }
        }
    };
}
test('Test restartConnection', (t) => {
    const _setup = sinon_1.stub().returns(Promise.resolve());
    const opts = mockOpts();
    const _setTimeout = sinon_1.spy();
    const url = '';
    const exchange = '';
    const result = index_2.restartConnection({ opts: opts.mocks, _setup, _setTimeout });
    t.ok(result instanceof Promise, 'RestartConnection returns a promise');
    t.ok(_setTimeout.calledOnce, 'Set a timeout');
    t.notOk(opts.mocks._restartConnection.called, 'Do not iterate before timeout');
    t.notOk(_setup.called, 'Do not run setup before timeout');
    const timeoutCb = _setTimeout.getCall(0).args[0];
    timeoutCb();
    t.ok(_setup.calledOnce, 'Run setup un timeout');
    t.notOk(opts.mocks._restartConnection.called, 'Do not iterate if setup succeds');
    _setup.reset();
    _setup.returns(Promise.reject({}));
    timeoutCb();
    setTimeout(() => {
        t.ok(opts.mocks._restartConnection.calledOnce, 'Iterate if setup blows up on timeout');
        t.end();
    }, 0);
});
test('Tests setup funcion', (t) => {
    const opts = mockOpts();
    function test() {
        return __awaiter(this, void 0, void 0, function* () {
            const result = yield index_1.default(opts.mocks);
            t.ok(opts.steps.register.calledOnce, 'Add is being call for each one of the provided registrations');
            const passAddition = opts.steps.register.getCall(0).args[0];
            t.deepEqual(opts.mocks.registrations.test, passAddition, 'The subscription passed to register is the expected one');
            opts.steps.register.reset();
            yield result.register({ pattern: '', handler: sinon_1.spy() });
            t.ok(opts.steps.register.calledOnce, 'Returns an initialized register function');
            yield result.request({ pattern: '', payload: Buffer.from('{}') });
            t.ok(opts.steps.request.calledOnce, 'Returns an initialized act function');
            const on0 = opts.steps.connectResponse.on.getCall(0);
            const on1 = opts.steps.connectResponse.on.getCall(1);
            t.equals(on0 && on0.args[0], 'error', 'It adds a handler to connection error');
            t.equals(on1 && on1.args[0], 'close', 'It adds a handler to connection close');
            const register = sinon_1.spy();
            const request = sinon_1.spy();
            opts.mocks._restartConnection.returns(Promise.resolve({ request, register }));
            on0.args[1]();
            t.ok(opts.mocks._restartConnection.calledOnce, 'Restart connection is called on connection close');
            yield wait(0); // Wait for the connection to stablish again;
            yield result.register({ pattern: '', handler: sinon_1.spy() });
            t.ok(register.calledOnce, 'After successsfull restart register is reasigned');
            yield result.request({ pattern: '', payload: Buffer.from('') });
            t.ok(request.calledOnce, 'After successsfull restart act is reasigned');
            opts.mocks._restartConnection.returns(Promise.reject({}));
            on0.args[1]();
            yield result.register({ pattern: '', handler: sinon_1.spy() }).then(() => {
                t.ok(true, 'Subscribe works on errored library');
            });
            yield result.request({ pattern: '', payload: Buffer.alloc(0) })
                .then(() => {
                t.ok(false, 'Emit should fail on errored library');
            })
                .catch(err => {
                t.ok(true, 'Emit rejects the promise if the pipe is broken');
            });
        });
    }
    test()
        .then(() => t.end())
        .catch(console.error);
});
