// @repositive/iris/dist/amqp/request.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const uuid_1 = require("uuid");
class TimeoutError extends Error {
    constructor(msg) {
        super(msg || 'Timeout');
    }
}
exports.TimeoutError = TimeoutError;
class RPCError extends Error {
    constructor(msg) {
        super(msg || 'Remote rejection');
    }
}
exports.RPCError = RPCError;
function setupRequest({ ch, exchange, _setTimeout = setTimeout, _clearTimeout = clearTimeout, _log = console }) {
    return __awaiter(this, void 0, void 0, function* () {
        return function request({ pattern, payload, timeout = 100 }) {
            return __awaiter(this, void 0, void 0, function* () {
                const q = yield ch.assertQueue('', { exclusive: true });
                const correlation = uuid_1.v4();
                const content = payload ? payload : Buffer.alloc(0);
                ch.publish(exchange, pattern, content, {
                    correlationId: correlation,
                    replyTo: q.queue
                });
                return new Promise((resolve, reject) => {
                    const time = _setTimeout(() => {
                        ch.deleteQueue(q.queue)
                            .then(() => {
                            reject(new TimeoutError('Timeout'));
                        })
                            .catch(reject);
                    }, timeout);
                    ch.consume(q.queue, (msg) => {
                        if (msg && msg.properties.correlationId === correlation) {
                            try {
                                _clearTimeout(time);
                                ch.deleteQueue(q.queue);
                                if (msg.properties.headers.code === 0) {
                                    resolve(msg.content);
                                }
                                else {
                                    reject(new RPCError(msg.content.toString()));
                                }
                            }
                            catch (err) {
                                reject(err);
                            }
                            ch.ack(msg);
                        }
                        //TODO: If the correlationId does not mach... We should never get here. If that's the case maybe move the msg to error queue?
                    });
                });
            });
        };
    });
}
exports.setupRequest = setupRequest;
