// @repositive/iris/dist/amqp/request.spec.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const test = require("tape");
const sinon_1 = require("sinon");
const request_1 = require("./request");
function mockChannel() {
    return {
        assertQueue: sinon_1.stub().returns({ queue: 'test' }),
        consume: sinon_1.spy(),
        sendToQueue: sinon_1.spy(),
        ack: sinon_1.spy(),
        publish: sinon_1.spy(),
        deleteQueue: sinon_1.stub().returns(Promise.resolve())
    };
}
function mockSerialization() {
    return {
        parse: sinon_1.stub().returns({}),
        serialize: sinon_1.stub().returns(Buffer.from('{}'))
    };
}
function wait(time) {
    return new Promise((resolve, reject) => {
        setTimeout(() => resolve(), time);
    });
}
test('Test request', (t) => {
    const ch = mockChannel();
    const exchange = '';
    function test() {
        return __awaiter(this, void 0, void 0, function* () {
            const pSetupRequest = request_1.setupRequest({ ch, exchange });
            t.ok(pSetupRequest instanceof Promise, 'Setup returns a promise');
            const request = yield pSetupRequest;
            const pattern = '';
            const payload = Buffer.from('');
            const pResult1 = request({ pattern, payload });
            t.ok(pResult1 instanceof Promise, 'Act returns a promise');
            yield pResult1
                .then(() => {
                t.notOk(true, 'Should not return if there was no response');
            })
                .catch(err => {
                t.ok(ch.deleteQueue.calledOnce, 'Deletes queue on timeout');
                t.equals(err && err.message, 'Timeout', 'Throws timeout if there is no response');
            });
            ch.publish.reset();
            ch.deleteQueue.reset();
            ch.consume.reset();
            const pResult2 = request({ pattern, payload });
            yield wait(0);
            const pCall = ch.publish.getCall(0);
            t.deepEquals(ch.publish.calledOnce && pCall.args[2], payload, 'Publishes the payload');
            const cCall = ch.consume.getCall(0);
            t.equals(ch.consume.calledOnce && ch.consume.getCall(0).args[0], 'test', 'Consumes the queue');
            const r = Math.random();
            const content = Buffer.from(JSON.stringify({ r }));
            const properties = pCall.args[3];
            properties.headers = { code: 0 };
            ch.consume.getCall(0).args[1]({ content, properties });
            yield pResult2
                .then((result) => {
                t.deepEquals(Buffer.from(JSON.stringify({ r })), result, 'On success get the expected result');
                t.ok(ch.deleteQueue.calledOnce, 'Deletes the queue on message received');
                t.ok(ch.ack.calledOnce, 'Acknowledges the message reception');
            })
                .catch((err) => {
                t.notOk(true, 'On success it should not reject');
            });
            ch.publish.reset();
            ch.deleteQueue.reset();
            ch.deleteQueue.returns(Promise.resolve());
            ch.consume.reset();
            ch.ack.reset();
            const pResult3 = request({ pattern, payload });
            yield wait(0);
            const errContent = Buffer.from(JSON.stringify({ r }));
            const pCall3 = ch.publish.getCall(0);
            const prop3 = pCall3.args[3];
            prop3.headers = { code: 1 };
            ch.consume.getCall(0).args[1]({ content: errContent, properties: prop3 });
            yield pResult3
                .then((result) => {
                t.notOk(true, 'On error can not execute then');
            })
                .catch((err) => {
                t.ok(err instanceof request_1.RPCError, 'The error is an instance of RPCError');
                t.deepEquals(err.message, JSON.stringify({ r }), 'On error get the expected error message');
                t.ok(ch.deleteQueue.calledOnce, 'Deletes the queue on message on message rejected');
                t.ok(ch.ack.calledOnce, 'Acknowledges the message reception on message rejection');
            });
        });
    }
    test()
        .then(() => t.end())
        .catch(console.error);
});
