// @repositive/iris/dist/examples/chat.example.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const __1 = require("..");
const readLine = require('readline');
const rl = readLine.createInterface({
    input: process.stdin,
    output: process.stdout
});
function question({ query, _rl = rl }) {
    return new Promise((resolve, reject) => {
        _rl.question(query, (answer) => {
            resolve(answer);
        });
    });
}
function chatListener({ payload }) {
    return __awaiter(this, void 0, void 0, function* () {
        console.log(`\n${payload.author}: ${payload.comment}`);
        return { ack: true };
    });
}
function prepareNameListener(username) {
    return function nameListener({ payload }) {
        return __awaiter(this, void 0, void 0, function* () {
            return {
                name: username
            };
        });
    };
}
const cmds = {
    'cntg': ({ oldParams, _question = question }) => __awaiter(this, void 0, void 0, function* () {
        const target = yield _question({ query: 'New target: ' });
        return Object.assign({}, oldParams, { target });
    })
};
function chat({ request, username, target, _question = question, _cmds = cmds }) {
    return __awaiter(this, void 0, void 0, function* () {
        while (true) {
            const answer = yield _question({ query: `${username}: ` });
            if (/:[a-z]/.test(answer)) {
                const cmd = answer.substring(1);
                return chat(yield _cmds[cmd]({ oldParams: { request, username, target, _question } }));
            }
            else {
                yield request({ pattern: `chat.${target}`, payload: { comment: `${answer}`, author: username } }).catch(console.warn);
            }
        }
    });
}
function init({ /* Default argument definition */ _question = question, _irisSetup = __1.default }) {
    return __awaiter(this, void 0, void 0, function* () {
        const { request, register } = yield _irisSetup({ namespace: 'chat' });
        const username = yield _question({ query: 'Your username: ' });
        register({ pattern: `chat.${username}`, handler: chatListener });
        register({ pattern: `chat.name.${username}`, handler: prepareNameListener(username) });
        const target = yield _question({ query: 'Insert user: ' });
        return chat({ request, username, target });
    });
}
init({}).catch(console.error);
